app/src/main/java/com/lucas/medicaltools/MainActivity.kt

```diff
@@ -1,11 +1,76 @@
 package com.lucas.medicaltools
 
+import android.annotation.SuppressLint
+import android.app.SearchManager
+import android.content.Context
 import androidx.appcompat.app.AppCompatActivity
 import android.os.Bundle
+import android.view.Menu
+import android.view.View
+import android.widget.ProgressBar
+import androidx.appcompat.widget.SearchView
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+import com.lucas.medical_equip.repository.MedicalTool
+import com.lucas.medical_equip.service.RetrofitBuilder
+import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
+import io.reactivex.rxjava3.schedulers.Schedulers
+import mu.KotlinLogging
 
+private val logger = KotlinLogging.logger {}
 class MainActivity : AppCompatActivity() {
+    private lateinit var medicalTools: List<MedicalTool>
+    private lateinit var medicalRecyclerView: RecyclerView
+    private lateinit var medicalAdapter: MedicalToolAdapter
+    private lateinit var progressBar: ProgressBar
+    private lateinit var searchView: SearchView
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        progressBar = findViewById(R.id.progressBar1)
+        getMedicalTools()
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
+        menuInflater.inflate(R.menu.menu_main, menu)
+        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
+        if (menu != null) {
+            searchView = menu.findItem(R.id.action_search).actionView as SearchView
+        }
+        searchView.setSearchableInfo(searchManager.getSearchableInfo(componentName))
+        searchView.maxWidth = Integer.MAX_VALUE
+        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
+            override fun onQueryTextSubmit(query: String?): Boolean {
+                return false
+            }
+
+            override fun onQueryTextChange(newText: String): Boolean {
+                medicalAdapter.filter.filter(newText)
```

Comment:
A lot of the time, we don't have the luxury of all the data sitting on the app. Imagine you had 10 000 entries and filtered that on every single text change, it would cause the app to hang.

My advise to you is to use Rx Java to enable filtering in the background and drop the filtering on the adapter. 

So what I want is for you to maybe use the cache'd repo to do filtering and it would do all that filtering in the background on each key press. 

What will be important is that on each key press, if you don't dispose the old observable, you end up in a situation where you actually have multiple items in the background, so just figure out how to dispose the old observable.